### ISSUE TITLE
How to Convert Python to C# in this code?

### ISSUE THREAD

**Issue by komjun** [migueldeicaza/TensorFlowSharp#311]
x = np.array([[0, 0], [1, 0], [1, 1], [0, 1]], np.float32) # 4x2, input
y = np.array([0, 0, 1, 0], np.float32) # 4, correct output, AND operation
#y = np.array([0, 1, 1, 1], np.float32) # OR operation
y = np.reshape(y, [4,1]) # convert to 4x1

X = tf.placeholder(tf.float32, shape=[4, 2])
Y = tf.placeholder(tf.float32, shape=[4, 1])

W = tf.Variable(tf.zeros([NUM_FEATURES, 1]), tf.float32)
B = tf.Variable(tf.zeros([1, 1]), tf.float32)

yHat = tf.sigmoid( tf.add(tf.matmul(X, W), B) ) # 4x1
err = Y - yHat <--- This!
deltaW = tf.matmul(tf.transpose(X), err ) # have to be 2x1
deltaB = tf.reduce_sum(err, 0) # 4, have to 1x1. sum all the biases? yes
W_ = W + learning_rate * deltaW
B_ = B + learning_rate * deltaB

step = tf.group(W.assign(W_), B.assign(B_)) #to update the values of weights and biases.

sess = tf.Session()
init = tf.global_variables_initializer()
sess.run(init)

for k in range(NUM_ITER):
 sess.run([step], feed_dict={X: x, Y: y})

W = np.squeeze(sess.run(W))
b = np.squeeze(sess.run(B))

plot_x = np.array([np.min(x[:, 0] - 0.2), np.max(x[:, 1]+0.2)])
plot_y = - 1 / W[1] * (W[0] * plot_x + b)
plot_y = np.reshape(plot_y, [2, -1])
plot_y = np.squeeze(plot_y)

print('W: ' + str(W))
print('b: ' + str(b))
print('plot_y: '+ str(plot_y))

plt.scatter(x[:, 0], x[:, 1], c=y, s=100, cmap='viridis')
plt.plot(plot_x, plot_y, color='k', linewidth=2)
plt.xlim([-0.2, 1.2]); plt.ylim([-0.2, 1.25]);
plt.show()

How do i write C# code ?
i try this

using (var session = new TFSession())
            {
                var graph = session.Graph;
                float[,] x = new float[4,2] { { 0, 0 }, { 0, 1 }, { 1, 0 }, { 1, 1 } };
                float[] y = new float[4] { 0,0,0,1 };
                var X = graph.Placeholder(TFDataType.Float, new TFShape(4, 2));
                var Y = graph.Placeholder(TFDataType.Float, new TFShape(4, 1));
                var W = graph.Variable(graph.Zeros(new TFShape(2, 1), TFDataType.Float));
                var B = graph.Variable(graph.Zeros(new TFShape(1, 1), TFDataType.Float));
                var yHat = graph.Sigmoid(graph.Add(graph.MatMul(X,W),B));
                var err = Y - yHat; <-- Error!
                var deltaW = graph.MatMul(graph.Transpose(X),err);
            }
            Console.Read();

**Comment by DinoSourcesRex** [comment]
I think it would be good to have a basic example in python code converted to TensorflowSharp as well.

**Comment by Andy-Wilkinson** [comment]
Since TensorFlowSharp uses explicit graph generation (unlike Python that has a global graph) you have to use the TFGraph.Sub(...) method rather than the subtraction operator. Note that this applies to all operators.

So replace the line,
`var err = Y - yHat;`

with,
`var err = graph.Sub(Y, yHat);`
